lib/vsc/filesystem/oceanstor.py

```diff
@@ -0,0 +1,148 @@
+        super(OceanStorClient, self).__init__(*args, **kwargs)
+
+        # X-Auth-Token header
+        self.x_auth_header = None
+
+        if ssl_verify is False:
+            # Disable verification of SSL certificates
+            nossl_context = ssl._create_unverified_context()
+            nosslHandler = HTTPSHandler(context=nossl_context)
+            self.opener = build_opener(nosslHandler)
+            fancylogger.getLogger().warning("Verification of SSL certificates disabled by request!")
+
+    def request(self, *args, **kwargs):
+        """
+        Wrapper for Client.request() with HTTP error and exit code handling
+        Injects X-Auth-Token headers into teh query if present
+        """
+        # Inject X-Auth-Token into headers (args=(method, url, body, headers))
+        args = list(args)
+        if args[3] is None:
+            args[3] = {}
+        if self.x_auth_header:
+            args[3].update(self.x_auth_header)
+
+        # Execute request catching any HTTPerror
+        try:
+            status, response = super(OceanStorClient, self).request(*args, **kwargs)
+        except HTTPError as err:
+            errmsg = "OceanStor query failed with HTTP error: %s (%s)" % (err.reason, err.code)
+            fancylogger.getLogger().error(errmsg)
+            raise
+
+        # Query exit code
+        try:
+            result = response['result']
+        except AttributeError as err:
+            errmsg = "OceanStor response lacks resulting status: %s" % response
+            fancylogger.getLogger().raiseException(errmsg, exception=AttributeError)
+        else:
+            ecmsg = "OceanStor query returned exit code: %s (%s)" % (result['code'], result['description'])
+            if result['code'] != 0:
+                fancylogger.getLogger().raiseException(ecmsg, exception=RuntimeError)
+            else:
+                fancylogger.getLogger().debug(ecmsg)
+
+        return status, response
+
+    def get_x_auth_token(self, password, username=None):
+        """Request authetication token"""
+        if not username:
+            username = self.username
+
+        query_url = os.path.join('aa', 'sessions')
+        payload = {
+            'user_name': username,
+            'password': password,
+        }
+
+        status, response = self.post(query_url, body=payload)
+        print(status, response)
+        token = response['data']['x_auth_token']
+        self.x_auth_header = {'X-Auth-Token': token}
+        fancylogger.getLogger().info("OceanStor authentication switched to X-Auth-Token for this session")
+
+        return True
+
+
+class OceanStorRestClient(RestClient):
+    def __init__(self, *args, **kwargs):
+        """Create client for OceanStor with given arguments"""
+        self.client = OceanStorClient(*args, **kwargs)
+
+
+class OceanStorOperations(with_metaclass(Singleton, PosixOperations)):
+    def __init__(self, username=None, password=None):
```

Comment:
can we make the url also a argument here? There are 2 instances of this storage system

Reply:
done in https://github.com/vub-hpc/vsc-filesystem-oceanstor/pull/1/commits/1920aaec0d381e99c62f6a50f3558ef6be1fad86